mqttclients/dotnet/MQTTnet.Client.Extensions/X509ChainValidator.cs

```diff
@@ -1,33 +1,70 @@
 ﻿using System.Diagnostics;
+using System.Net.Security;
 using System.Security.Cryptography.X509Certificates;
 
 namespace MQTTnet.Client.Extensions
 {
     internal static class X509ChainValidator
     {
-        internal static bool ValidateChain(X509Certificate cert, X509Certificate2Collection? certChain = null)
+
+        internal static bool ValidateChain(MqttClientCertificateValidationEventArgs certValArgs, string caCertFile = "")
         {
-            X509Chain chain = new();
-            chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
-            chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
-            chain.ChainPolicy.VerificationFlags = X509VerificationFlags.NoFlag;
-            chain.ChainPolicy.VerificationTime = DateTime.Now;
-            chain.ChainPolicy.UrlRetrievalTimeout = new TimeSpan(0, 0, 0);
-            if (certChain != null)
+            X509Certificate2Collection caCerts = new();
+            if (!string.IsNullOrEmpty(caCertFile))
             {
-                chain.ChainPolicy.CustomTrustStore.AddRange(certChain);
-                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+                caCerts.ImportFromPemFile(caCertFile);
             }
-            X509Certificate2 x5092 = new X509Certificate2(cert);
-            bool res = chain.Build(x5092);
-            if (res == false)
+            return ValidateChain(certValArgs, caCerts);
+        }
+
+        internal static bool ValidateChain(MqttClientCertificateValidationEventArgs certValArgs, X509Certificate2Collection caChain)
+        {
+            if (certValArgs.SslPolicyErrors == SslPolicyErrors.None)
             {
-                Trace.TraceError($"Error validating TLS chain for cert: '{cert.Subject}' issued by '{cert.Issuer}'");
-                chain.ChainStatus.ToList().ForEach(s => Trace.TraceError(s.StatusInformation));
+                return true;
             }
-            return res;
-        }
 
+            Trace.TraceError($"SslPolicyErrors: '{certValArgs.SslPolicyErrors}'");
+            Trace.TraceWarning($"Loaded {caChain.Count} certs ");
+            caChain.ToList().ForEach(c => Trace.TraceWarning(c.Subject));
+
+            var sslErrors = Enum.GetValues<SslPolicyErrors>().Where(e => certValArgs.SslPolicyErrors.HasFlag(e)).ToList();
+            bool valid = false;
 
+            foreach (var sslError in sslErrors)
+            {
+                if (sslError == SslPolicyErrors.RemoteCertificateChainErrors)
+                {
+                    bool chainValidated = false;
+                    X509Chain chain = new();
```

Comment:
Is there any way to assign these fields when `chain` is instantiated via object initializers?

Reply:
no, afaik X509Chain does not have initializers